### PR TITLE
Sn 5140 filter gsv

### PR DESCRIPTION
**Pull request by tmcd249**
NMEA protocol now supports filtering of GxGSV by constellation and frequency.

### CODE REVIEW

**Review by waltjohnson — APPROVED** [review]

**Review by kylemallory — CHANGES_REQUESTED** [review]

File: src/data_sets.h [hunk 1]
```diff
@@ -1849,7 +1851,16 @@ typedef struct PACKED
     /** IMU and Integrated IMU data transmit period is set using DID_SYS_PARAMS.navPeriodMs */
 } rmc_t;
 
+#define NMEA_GNGSV_FREQ_1_BIT    (0x01)
+#define NMEA_GNGSV_FREQ_2_BIT    (0x01 << 1)
+#define NMEA_GNGSV_FREQ_3_BIT    (0x01 << 2)
+#define NMEA_GNGSV_FREQ_5_BIT    (0x01 << 3)
 
+#define NMEA_GNGSV_GPS_OFFSET    (SAT_SV_GNSS_ID_GPS << 4)
```
[the remark this answers is not in the document]
Comment (tmcd249) [reply src/data_sets.h:1859]: this is an offset of (CONST_ID << 4) + NMEA_MSG_ID_SPECIAL_CASE_START * BASE_MSG_NUM

File: src/protocol_nmea.cpp [hunk 2]
```diff
@@ -13,45 +13,52 @@ static uint8_t s_gnssId = SAT_SV_GNSS_ID_GNSS;
...
+static gsvMask_t s_gsvMask = {0};
 
 //////////////////////////////////////////////////////////////////////////
 // Utility functions
 //////////////////////////////////////////////////////////////////////////
 
 void nmea_set_protocol_version(int protocol_version)
 { 
-	s_protocol_version = protocol_version; 
+    s_protocol_version = protocol_version;
```
[the remark this answers is not in the document]
Comment (tmcd249) [reply src/protocol_nmea.cpp:37]: I think these were added to work with the unit tests but honestly i don't know the reasoning

File: src/data_sets.h [hunk 3]
```diff
@@ -1849,7 +1851,16 @@ typedef struct PACKED
     /** IMU and Integrated IMU data transmit period is set using DID_SYS_PARAMS.navPeriodMs */
 } rmc_t;
 
+#define NMEA_GNGSV_FREQ_1_BIT    (0x01)
+#define NMEA_GNGSV_FREQ_2_BIT    (0x01 << 1)
+#define NMEA_GNGSV_FREQ_3_BIT    (0x01 << 2)
+#define NMEA_GNGSV_FREQ_5_BIT    (0x01 << 3)
```
[the remark this answers is not in the document]
Comment (tmcd249) [reply src/data_sets.h:1857]: There is no frequency 4 also. L3 is a Glonass, beidou and QZSS thing. 1,2,3,5 are the supported frequencys channels. L is only GPS and SBS.

File: src/protocol_nmea.cpp [hunk 4]
```diff
@@ -388,1189 +395,1365 @@ void nmea_enable_stream(uint32_t& bits, uint8_t* period, uint32_t nmeaId, uint8_
...
+    }
 
-	return n;
+    return n;
+}
+
+/**
+ * decodes the NMEA GSV family of messages
+ * Returns:
+ * 	-2 for invalid length
```
[the remark this answers is not in the document]
Comment (tmcd249) [reply src/protocol_nmea.cpp:1656]: fixed

File: src/protocol_nmea.cpp [hunk 5]
```diff
@@ -1623,375 +1806,433 @@ int nmeaMsgIdToTalker(int msgId, void *str, int strSize)
...
+    ptr = ASCII_to_f64(&(gpsPos.towOffset), ptr);
+    ptr = ASCII_to_u8(&(gpsPos.leapS), ptr);
 
-	return 0;
+    return 0;
+}
+
+/**
+ * Sets ASCE special case.
+ * returns 15 if successful
```
[the remark this answers is not in the document]
Comment (tmcd249) [reply src/protocol_nmea.cpp:2014]: fixed

File: src/protocol_nmea.cpp [hunk 6]
```diff
@@ -388,1189 +395,1365 @@ void nmea_enable_stream(uint32_t& bits, uint8_t* period, uint32_t nmeaId, uint8_
...
+int decodeGSV(char* a, int aSize)
+{
+    if(aSize < 6 || !(a))     // five characters required (i.e. "$INFO")
+        return -2;
+
+    int msgNum;
+    
+    if(a[1] == 'x')        return NMEA_MSG_ID_GxGSV;
+    else if (a[1] == 'N')  msgNum = NMEA_GNGSV_START;
+    else if (a[1] == 'P')  msgNum = NMEA_GNGSV_START | NMEA_GNGSV_GPS_OFFSET;
```
[the remark this answers is not in the document]
Comment (tmcd249) [reply src/protocol_nmea.cpp:1668]: it is and offset from the base number of NMEA_GNGSV_START. All GPS messages are offset 16 messages from their GNGSV counterpart. You are right about intialising. Adding that in.

[the remark this answers is not in the document]
Comment (kylemallory) [reply src/protocol_nmea.cpp:1668]: If it is an offset from GNGSV_START, then should it be msgNUm = NGMA_GNGSV_START + NMEA_GNGSV_x_OFFSET?   That would make more sense here.

File: src/protocol_nmea.cpp [hunk 7]
```diff
@@ -13,45 +13,52 @@ static uint8_t s_gnssId = SAT_SV_GNSS_ID_GNSS;
...
+    float			speed2dMps;
+    float			speed2dKnots;
 } s_dataSpeed;
 
 uint8_t nmea2p3_svid_to_sigId(uint8_t gnssId, uint16_t svId);
+bool gsv_freq_ena(gps_sig_sv_t* sig);
+
+typedef struct {
+    uint8_t constMask[SAT_SV_GNSS_ID_COUNT]; /* Constilation mask (see eGnGSVIndex)*/
+}gsvMask_t;
```
[the remark this answers is not in the document]
Comment (tmcd249) [reply src/protocol_nmea.cpp]: agreed

File: src/protocol_nmea.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (tmcd249) [reply src/protocol_nmea.cpp]: Fixed

**Review by kylemallory — APPROVED** [review]

**Pull request merged**